oracle/config/half-duplex-transfer-watcher.config.js

```diff
@@ -0,0 +1,38 @@
+const baseConfig = require('./base.config')
+const { ERC20_ABI } = require('../../commons')
+const { EXIT_CODES } = require('../src/utils/constants')
+
+const initialChecksJson = process.argv[3]
+
+if (!initialChecksJson) {
+  throw new Error('initial check parameter was not provided.')
+}
+
+let initialChecks
+try {
+  initialChecks = JSON.parse(initialChecksJson)
+} catch (e) {
+  throw new Error('Error on decoding values from initial checks.')
+}
+
+const id = `${baseConfig.id}-half-duplex-transfer`
+
+const transferWatcherRequired = baseConfig.id === 'erc-native'
+
+if (!transferWatcherRequired) {
+  console.error(`Transfer watcher not required for bridge mode ${process.env.ORACLE_BRIDGE_MODE}`)
+  process.exit(EXIT_CODES.WATCHER_NOT_REQUIRED)
+}
+
+module.exports = {
+  ...baseConfig.bridgeConfig,
+  ...baseConfig.foreignConfig,
+  event: 'Transfer',
+  eventContractAddress: initialChecks.halfDuplexTokenAddress,
```

Comment:
we need to consider the case when `halfDuplexTokenAddress` is not defined. There are two possible scenarios:
- the oracle has been updated before the upgrade of the bridge contracts happens
- the oracle is used with a token different from Dai

Reply:
I'm trying to look for approaches for this. At the moment of starting the halfDuplexToken watcher we could find the following scenarios:
- `halfDuplexErc20token` method does not exist (contract were not upgraded yet or is not dai).
- `halfDuplexErc20token` method exists but it returns the same address as the `erc20token` method (contracts were updated but `migrateToMCD` wasn't invoked).

In both scenarios, halfDuplexToken watcher and the swap tokens worker should not perform any actions. The swap tokens worker is subscribed to a queue, so it will be idle while no watcher pushes a job to that queue.

Regarding the halfDuplexToken watcher, we can end the process after checking the existence of the methods but docker will try to restart them every time and keep the watcher in a loop of starting the app, performing checks, ending the app.

We can use a different docker-compose file if the conditions are not met to avoid including the new watcher and worker, but after performing the contract upgrade and migration, it will require more than just restarting the oracle component.  

One idea could be to add a functionality that a worker could be set to sleep forever if some conditions are met.

For example, on the initial checks when the watcher is starting, if one of the two scenarios are detected, we can set an attribute and then set the watcher to sleep instead of listening to the events.

So after the contracts are upgraded and `migrateToMCD` is invoked, restarting the components will make the watcher start working correctly.

Reply:
thanks for analysis! Did you consider a way to modify [the `main` function](https://github.com/poanetwork/tokenbridge/blob/ab2c0ea120fa1efe6af213ce9400b7cb0a291bee/oracle/src/watcher.js#L131) as so it check if the watcher is for the half duplex token transfers and if the both `halfDuplexErc20token()` and `erc20token()` return the same address (or `halfDuplexErc20token()` fails) it does `return`?

Reply:
Thanks for the suggestion! Please check https://github.com/poanetwork/tokenbridge/pull/244/commits/8fa715089b5a4e6e142b28c46990cf940f91cc27

Reply:
Awesome! Does it mean that we need to restart the oracle after the `migrateToMCD` call?

Reply:
That's correct. After calling `migrateToMCD` we need to restart the oracle. After the restart, the `Transfer-watcher` will start listening to events of the updated `erc20token()` address and the `halfDuplexToken-transfer-watcher` will start to listen to events of the `halfDuplexErc20token()` address

Reply:
It is always a pain to synchronize validators to perform some actions. That's why do you think it is a bad idea to move

https://github.com/poanetwork/tokenbridge/blob/8fa715089b5a4e6e142b28c46990cf940f91cc27/oracle/scripts/initialChecks.js#L20-L30

to [the main function](https://github.com/poanetwork/tokenbridge/blob/8fa715089b5a4e6e142b28c46990cf940f91cc27/oracle/src/watcher.js#L133)?

Reply:
This will give the oracle some flexibility which will be good to avoid restarting the services. The only bad side is that it will add some overhead (RPC requests to get addresses) for the specific watchers. I'll update the code to support running checks before getting the events and update the variables needed

Reply:
Updated the implementation and also added an e2e test case to cover that the oracle works before and after the migration. Please check https://github.com/poanetwork/tokenbridge/pull/244/commits/3cd53f7bdab5a4f203ee0d8a94ba848c5e9a0f78